client/token.go

```diff
@@ -0,0 +1,474 @@
+type TokenClient interface {
+	CreateCollection(ctx context.Context, creator models.SingleSigner, req CreateCollectionRequest) (string, error)
+	CreateToken(ctx context.Context, creator models.SingleSigner, req CreateTokenRequest) (string, error)
+	MintToken(ctx context.Context, minter models.SingleSigner, req MintTokenRequest) (string, error)
+	OfferToken(ctx context.Context, sender models.SingleSigner, req OfferTokenRequest) (string, error)
+	ClaimToken(ctx context.Context, receiver models.SingleSigner, req ClaimTokenRequest) (string, error)
+
+	GetCollectionData(ctx context.Context, creator models.AccountAddress, collectionName string) (*CollectionData, error)
+	GetTokenData(ctx context.Context, creator models.AccountAddress, collectionName, tokenName string) (*TokenData, error)
+	GetToken(ctx context.Context, owner models.AccountAddress, tokenID TokenID) (*Token, error)
+}
+
+// NewTokenClient creates TokenClient to do things with aptos token.
+func NewTokenClient(client AptosClient) (TokenClient, error) {
+	ledgerInfo, err := client.LedgerInformation(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("get ledger info error: %v", err)
+	}
+
+	return &TokenClientImpl{
+		client:  client,
+		chainID: ledgerInfo.ChainID,
+	}, nil
+}
+
+type TokenClientImpl struct {
+	client  AptosClient
+	chainID uint8
+}
+
+var DefaultMaxGasAmount uint64 = 1000
+
+var DefaultMaxium uint64 = 1<<64 - 1
+
+var TokenModule models.Module
+var TokenTransferModule models.Module
+
+func init() {
+	moduleAddr, _ := models.HexToAccountAddress("0x3")
+	TokenModule = models.Module{
+		Address: moduleAddr,
+		Name:    "token",
+	}
+	TokenTransferModule = models.Module{
+		Address: moduleAddr,
+		Name:    "token_transfers",
+	}
+}
+
+type CreateCollectionRequest struct {
+	Name         string
+	Description  string
+	URI          string
+	Maximum      uint64
+	MutateConfig CollectionMutabilityConfig
+}
+
+type CollectionMutabilityConfig struct {
+	Description bool
+	URI         bool
+	Maximum     bool
+}
+
+func (impl *TokenClientImpl) CreateCollection(ctx context.Context, creator models.SingleSigner, req CreateCollectionRequest) (string, error) {
+	tx := models.Transaction{}
+
+	addr := creator.AccountAddress.PrefixZeroTrimmedHex()
+
+	accountInfo, err := impl.client.GetAccount(ctx, addr)
+	if err != nil {
+		return "", fmt.Errorf("get account info error: %v", err)
+	}
+
+	gasPrice, err := impl.client.EstimateGasPrice(ctx)
+	if err != nil {
+		return "", fmt.Errorf("get estimate gas price error: %v", err)
+	}
+
+	err = tx.SetChainID(impl.chainID).
+		SetSender(addr).
+		SetPayload(models.EntryFunctionPayload{
+			Module:   TokenModule,
+			Function: "create_collection_script",
+			Arguments: []interface{}{req.Name, req.Description, req.URI, req.Maximum,
+				[]bool{req.MutateConfig.Description, req.MutateConfig.Maximum, req.MutateConfig.URI}},
+		}).
+		SetExpirationTimestampSecs(uint64(time.Now().Add(30 * time.Second).Unix())).
+		SetGasUnitPrice(gasPrice).
```

Comment:
Referring to TypeScript SDK, it would estimate gas price if the caller doesn't provide one.
But it should be able to be decided by the caller as well. We can design similar options for `maxGasAmount`, `gasUnitPrice` and `expireTimestamp` later.
https://github.com/aptos-labs/aptos-core/blob/469f8b9138cdebf992689cb03b0164b79ccc29f9/ecosystem/typescript/sdk/src/aptos_client.ts#L657